diff -u -r -N arepo-master/Makefile arepo-agama/Makefile
--- arepo-master/Makefile	2024-03-20 11:05:32.000000000 +0000
+++ arepo-agama/Makefile	2024-03-20 11:27:03.000000000 +0000
@@ -320,6 +320,9 @@
 
 LIBS = $(GMP_LIB) $(MATH_LIB) $(MPICH_LIB) $(HDF5_LIB) $(GSL_LIB) $(FFTW_LIB) $(HWLOC_LIB)
 
+ifeq (EXTERNALGRAVITY_AGAMA, $(findstring EXTERNALGRAVITY_AGAMA, $(CONFIGVARS)))
+endif
+
 FOPTIONS = $(OPTIMIZE)
 FFLAGS = $(FOPTIONS)
 
diff -u -r -N arepo-master/Template-Config.sh arepo-agama/Template-Config.sh
--- arepo-master/Template-Config.sh	2023-04-29 14:33:30.000000000 +0100
+++ arepo-agama/Template-Config.sh	2024-03-20 11:23:30.000000000 +0000
@@ -89,6 +89,9 @@
 #EXTERNALGRAVITY               # master switch for external potential
 #EXTERNALGY=0.0                # constant external gravity in y direction
 
+#--------------------------------------- External potential provided by the Agama package
+#EXTERNALGRAVITY_AGAMA         # external potential provided by the Agama package and described by parameters in agama_potential.ini
+
 #--------------------------------------- Static NFW Potential
 #STATICNFW                     # static gravitational Navarro-Frenk-White (NFW) potential
 #NFW_C=12                      # concentration parameter of NFW potential
diff -u -r -N arepo-master/documentation/source/config-options.md arepo-agama/documentation/source/config-options.md
--- arepo-master/documentation/source/config-options.md	2023-04-29 14:33:30.000000000 +0100
+++ arepo-agama/documentation/source/config-options.md	2024-03-20 11:24:40.000000000 +0000
@@ -610,6 +610,13 @@
 
 Constant external gravity in the y-direction
 
+-------
+
+**EXTERNALGRAVITY_AGAMA**
+
+Activates an external potential provided by the Agama package,
+with parameters given in the file agama_potential.ini
+
 -----
 
 NFW Potential
diff -u -r -N arepo-master/src/gravity/grav_external.c arepo-agama/src/gravity/grav_external.c
--- arepo-master/src/gravity/grav_external.c	2023-04-29 14:33:30.000000000 +0100
+++ arepo-agama/src/gravity/grav_external.c	2024-03-20 11:20:59.000000000 +0000
@@ -47,6 +47,11 @@
 
 #include "../domain/domain.h"
 
+#ifdef EXTERNALGRAVITY_AGAMA
+void* agama_potential = NULL;
+extern double agama_evalPotential(const void* potential, const double pos[3], double time, double deriv[3], double deriv2[6]);
+#endif
+
 #ifdef EXTERNALGRAVITY
 static void gravity_external_get_force(double pos[3], int type, MyIDType ID, double acc[3], double *pot, int *flag_set);
 
@@ -71,14 +76,20 @@
       if(i < 0)
         continue;
 
-      double *pos;
+      double pos[3];
 
 #ifdef CELL_CENTER_GRAVITY
-      if(P[i].Type == 0)
-        pos = SphP[i].Center;
-      else
+      if(P[i].Type == 0) {
+        pos[0] = SphP[i].Center[0];
+        pos[1] = SphP[i].Center[1];
+        pos[2] = SphP[i].Center[2];
+      } else
 #endif /* #ifdef CELL_CENTER_GRAVITY */
-        pos = P[i].Pos;
+      {
+        pos[0] = P[i].Pos[0];
+        pos[1] = P[i].Pos[1];
+        pos[2] = P[i].Pos[2];
+      }
 
       double acc[3], pot;
       int flag_set = 0;
@@ -158,6 +169,16 @@
   *pot = -(EXTERNALGY)*pos[1];
 #endif /* #ifdef EXTERNALGY */
 
+#ifdef EXTERNALGRAVITY_AGAMA
+  {
+    double deriv[3];
+    *pot += agama_evalPotential(agama_potential, pos, All.Time, deriv, NULL);
+    acc[0] -= deriv[0];
+    acc[1] -= deriv[1];
+    acc[2] -= deriv[2];
+  }
+#endif
+
 #ifdef STATICISO
   {
     double r, m;
diff -u -r -N arepo-master/src/init/begrun.c arepo-agama/src/init/begrun.c
--- arepo-master/src/init/begrun.c	2023-04-29 14:33:30.000000000 +0100
+++ arepo-agama/src/init/begrun.c	2024-03-20 11:21:15.000000000 +0000
@@ -58,6 +58,12 @@
 herr_t my_hdf5_error_handler(void *unused);
 #endif
 
+#ifdef EXTERNALGRAVITY_AGAMA
+extern void* agama_potential;
+extern const char* agama_getError();
+extern void* agama_createPotential(const char* params);
+#endif
+
 static void delete_end_file(void);
 
 /*! \brief Prints a welcome message.
@@ -192,6 +198,17 @@
   All.FlushLast = CPUThisRun;
 #endif /* #ifdef REDUCE_FLUSH */
 
+#ifdef EXTERNALGRAVITY_AGAMA
+  char param[100] = "file=agama_potential.ini G=";
+  snprintf(param + strlen(param), 99 - strlen(param), "%.16g", All.G);
+  agama_potential = agama_createPotential(param);
+  if(!agama_potential) {
+    mpi_printf("BEGRUN: Cannot initialize Agama potential, halting.\n%s\n", agama_getError());
+    exit(1);
+  }
+  mpi_printf("BEGRUN: Initialized Agama potential with %s.\n", param);
+#endif
+
   init_scalars();
 
   init_gradients();
